onnxruntime/core/graph/graph.cc

```diff
@@ -2074,6 +2074,18 @@ Status Graph::InferAndVerifyTypeMatch(Node& node, const OpSchema& op, const Reso
     ORT_TRY {
       context.RunInferencing();
     }
+    ORT_CATCH(const onnx::InferenceError& ex) {
+      ORT_HANDLE_EXCEPTION([&]() {
+        // Relax shape inference check for If node and subgraph.
+        // ONNX shape inference may fail for a branch that won't get executed at runtime.
+        if (std::string(ex.what()).find("[ShapeInferenceError]") != std::string::npos) {
```

Comment:
>if (std::string(ex.what()).find("[ShapeInferenceError]" [](start = 8, length = 55)

This does seem to be a shaky conditional based on string matching. Can we add some enum checks?